optimism/src/trace.rs

```diff
@@ -0,0 +1,62 @@
+use crate::{lookups::Lookup, E};
+use ark_ff::Field;
+use kimchi_msm::witness::Witness;
+use std::{collections::HashMap, hash::Hash};
+
+pub struct Trace<const N: usize, SELECTOR, F> {
+    /// The domain size of the circuit
+    pub domain_size: usize,
+    /// The witness for a given selector
+    pub witness: HashMap<SELECTOR, Witness<N, Vec<F>>>,
+    /// The vector of constraints for a given selector
+    pub constraints: HashMap<SELECTOR, Vec<E<F>>>,
+    /// The vector of lookups for a given selector
+    pub lookups: HashMap<SELECTOR, Vec<Lookup<E<F>>>>,
+}
+
+impl<const N: usize, SELECTOR: Eq + Hash, F: Field> Trace<N, SELECTOR, F> {
+    /// Returns a boolean indicating whether the witness for the given selector is empty.
+    pub fn witness_is_empty(&self, step: SELECTOR) -> bool {
+        self.witness[&step].cols[0].is_empty()
+    }
+
+    /// Resets the witness after folding
+    pub fn reset(&mut self, step: SELECTOR) {
+        self.witness.insert(
+            step,
+            Witness {
+                cols: Box::new(std::array::from_fn(|_| {
+                    Vec::with_capacity(self.domain_size)
```

Comment:
Is there not a function to reset the vector instead of reallocating?

Reply:
Mmmm probably there's a `clear`, will check that out